feat(messages): add :!cmd shell message kinds

luukvbaal opened the pull request:
**feat(messages): add :!cmd shell message kinds**

Also print stderr error messages with ErrorMsg highlight group.

Fix #31901

**fix(messages): verbose kind for nvim_echo()**

Problem:  No "verbose" kind for nvim_echo() opts->verbose.
Solution: Pass NULL "kind" to indicate no new kind.

luukvbaal committed:
feat(messages): add :!cmd shell message kinds

Also print stderr error messages with ErrorMsg highlight group.

luukvbaal commented:
Not sure if this granularity is warranted, considering we [rejected](https://github.com/neovim/neovim/pull/31279#discussion_r1850566968) more specific kinds for various "list_cmd" output. Though those messages can still be parsed for known formatting whereas shell stderr/out/return can contain anything so the distinction may be useful. Otherwise just use "shell" for all chunks.

justinmk commented:
> Not sure if this granularity is warranted

No strong feelings either way. Perhaps an argument against the granularity is that `:!` will become not so special once we reimplement it as a PTY wrapper. In theory it won't be any different than `:echo system(...,{pty=true})`. Though it will continue to be able to handle stdout/stderr/exitcode, so these granular kinds can continue to be supported.

luukvbaal committed:
fix(messages): verbose kind for nvim_echo()

Problem:  No "verbose" kind for nvim_echo() opts->verbose.
Solution: Pass NULL "kind" to indicate no new kind.

zeertzjq commented on runtime/doc/news.txt:
```diff
@@ -96,7 +96,7 @@ EVENTS
     • `history` argument indicating if the message was added to the history.
     • new message kinds: "bufwrite", "completion", "list_cmd", "lua_print",
-      "search_cmd", "undo", "verbose", wildlist".
+      "search_cmd", "shell_out/err/return", "undo", "verbose", wildlist".
```
```suggestion
      "search_cmd", "shell_out/err/ret", "undo", "verbose", wildlist".
```

justinmk commented on src/nvim/event/proc.h:
```diff
@@ -21,8 +21,8 @@ static inline Proc proc_init(Loop *loop, ProcType type, void *data)
-    .out = { .s.closed = false },
-    .err = { .s.closed = false },
+    .out = { .s.closed = false, .s.fd = 1 },
+    .err = { .s.closed = false, .s.fd = 2 },
```
should we use `STDOUT_FILENO` / `STDERR_FILENO` ?

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -1067,7 +1068,7 @@ static void out_data_ring(const char *output, size_t size)
 
   if (output == NULL && size == SIZE_MAX) {   // Print mode
-    out_data_append_to_screen(last_skipped, &last_skipped_len, true);
+    out_data_append_to_screen(last_skipped, &last_skipped_len, 1, true);
```
STDOUT_FILENO ?
